tests/main/views/test_direct_award.py

```diff
@@ -934,22 +935,285 @@ def test_record_project_download_creates_audit_event(self):
         )
-        self.search_id = self.create_direct_award_project_search(created_by=self.user_id, project_id=self.project_id)
+        assert res.status_code == 404
+        assert json.loads(res.get_data()) == {
+            "error": "Project 976149527383380 not found",
+        }
+
+    @pytest.mark.parametrize(
+        (
+            "has_active_search",
+            "has_non_active_search",
+            "active_search_has_chosen_sid",
+            "non_active_search_has_chosen_sid",
+            "project_locked",
+            "existing_process_outcome_result",
+            "existing_process_outcome_complete",
+            "expected_status_code",
+            "expected_error_message",
+        ),
+        (
+            # "happy" paths
+            (True, True, True, True, True, None, False, 200, None,),
+            (True, False, True, False, True, None, False, 200, None,),
+            (True, False, True, False, True, "awarded", False, 200, None,),
+            (True, True, True, False, True, "cancelled", False, 200, None,),
+            # "failure" paths
+            (
+                True,
+                True,
+                False,
+                True,
+                True,
+                None,
+                False,
+                404,
+                AnyStringMatching(r"Service \d+ is not a result of project \d+\'s active search"),
+            ),
+            (
+                False,
+                True,
+                False,
+                True,
+                True,
+                None,
+                False,
+                404,
+                AnyStringMatching(r"Project \d+ doesn't have an active search"),
+            ),
+            (
+                True,
+                False,
+                True,
+                False,
+                False,
+                None,
+                False,
+                400,
+                AnyStringMatching(r"Project \d+ has not been locked"),
+            ),
+            (
+                True,
+                False,
+                False,
+                False,
+                False,
+                None,
+                False,
+                400,
+                AnyStringMatching(r"Project \d+ has not been locked"),
+            ),
+            (
+                True,
+                True,
+                False,
+                False,
+                True,
+                "cancelled",
+                True,
+                410,
+                AnyStringMatching(r"Project \d+ already has a completed process outcome: \d+"),
+            ),
+            (
+                True,
+                False,
+                True,
+                False,
+                True,
+                "awarded",
+                True,
+                410,
+                AnyStringMatching(r"Project \d+ already has a completed process outcome: \d+"),
+            ),
+        )
+    )
+    def test_most_paths(
```

Comment:
Once the test gets this name, maybe that's the point where we've taken parameterization too far ;-)

Reply:
What is the advantage of parametrising the test like this?

Reply:
It reduces repetition & copy-pastiness (which goes hand in hand with typos), and allows "expected conditions" -> "expected result" to be expressed as a simple tuple, making it extremely cheap to explore various combinations of options.

Reply:
This is very hard to read and understand. If we need this much data for a test then we need to find somewhere to explicitly store test data sets.

Reply:
So.. give the user the job of jumping back & forth between different files trying to piece together what argument corresponds to what, and then of course where would such a file go? We could discuss for hours. And then should it be json or python? How should it be imported or parsed?